src/Utils.ts

```diff
@@ -1,145 +1,133 @@
- *
- * @private
- * @param  {Array} array Array of something
- * @return {Array}       Shuffled array of something
- */
-export function shuffle<T>(array: T[]): T[] {
-  let currentIndex: number = array.length
-  let temporaryValue: T
-  let randomIndex: number
-
-  // While there remain elements to shuffle...
-  while (currentIndex !== 0) {
-    // Pick a remaining element...
-    randomIndex = Math.floor(Math.random() * currentIndex)
-    currentIndex -= 1
-    // And swap it with the current element.
-    temporaryValue = array[currentIndex]
-    array[currentIndex] = array[randomIndex]
-    array[randomIndex] = temporaryValue
-  }
-  return array
-}
-
-// Thanks mdn and damnit derpibooru
-/**
- * Generate a random int between [min, max]
- *
- * @private
- * @param {Number} min The minimum (inclusive)
- * @param {Number} max The maximum (inclusive)
- */
-export function randInt(min: number, max: number): number {
-  min = Math.ceil(min)
-  max = Math.floor(max)
-  return Math.floor(Math.random() * (max - min + 1)) + min
-}
-
-/**
- * Performs some basic search validation
- *
- * @private
- * @param {String} site The site to resolve
- * @param {Number|String} limit The limit for the amount of images to fetch
- */
-export function validateSearchParams(site: string, limit: number|string): {site: string, limit: number} {
-  const resolvedSite = resolveSite(site)
-
-  if (typeof limit !== 'number')
-    limit = parseInt(limit)
-
-  if (resolvedSite === null)
-    throw new BooruError('Site not supported')
-
-  if (typeof limit !== 'number' || Number.isNaN(limit))
-    throw new BooruError('`limit` should be an int')
-
-  return { site: resolvedSite, limit }
-}
-
-/**
- * Finds the matching strings between two arrays
- *
- * @private
- * @param {String[]} arr1 The first array
- * @param {String[]} arr2 The second array
- * @return {String[]} The shared strings between the arrays
- */
-export function compareArrays(arr1: string[], arr2: string[]): string[] {
-  const matches: string[] = []
-  arr1.forEach(ele1 => {
-    arr2.forEach(ele2 => {
-      if (ele1.toLowerCase() === ele2.toLowerCase()) {
-        matches.push(ele1)
-      }
-    })
-  })
-
-  return matches
-}
+/*
+ - Utils
+ => .resolveSite(site/alias)
+ => .jsonfy([images])
+ => .shuffle([arr])
+ => .randInt(min, max)
+ */
+
+import { parse as xml2json } from 'fast-xml-parser';
+import { BooruError, SiteMap as sites } from './Constants';
+
+/**
+ * Check if `site` is a supported site (and check if it's an alias and return the sites's true name)
+ *
+ * @param  {String} domain The site to resolveSite
+ * @return {String?} null if site is not supported, the site otherwise
+ */
+export const resolveSite = (domain: string): string | null => {
+  if (typeof domain !== 'string') {
+    return null;
+  }
+
+  domain = domain.toLowerCase();
+
+  for (const site in sites) {
+    if (site === domain || (sites[site].aliases as string[]).includes(domain)) {
```

Comment:
It is, generic typing is not involved here. Without casting it to a `string[]` it throws this: 
![image](https://user-images.githubusercontent.com/4019718/53306910-4c1fa100-3893-11e9-98bd-a90c3c1aa54b.png)

Reply:
Same. It worked because you had `SMap<any>` for your `booruCache` in `index.ts`.

Previously your SiteInfo had this:

https://github.com/AtlasTheBot/booru/blob/7f3c9ea169551920631be1bd1c59f69bf64f585b/src/structures/SiteInfo.ts#L13-L16

Which when specifying the type of `booruCache` I changed to:

```ts
 type?: string 
 /** The aliases of this site */ 
 aliases?: string[] 
 /** If this site serves NSFW posts or not */
```

Otherwise this was thrown:

![image](https://user-images.githubusercontent.com/4019718/53325621-03440880-38e4-11e9-8af5-b4e0f1c93dec.png)

So there were 3 solutions:

1. Mark `aliases` as optional in `SiteInfo` and cast it to `string[]` later while keeping the typings for booruCache
2. Remove typing for booruCache, keep as is
3. Hackily include all those parameters in the Booru class

I went for option 1.

Reply:
Changing that it is not required would have the following impact:

```diff
diff --git a/src/Constants.ts b/src/Constants.ts
index ad97fe3..2258b79 100644
--- a/src/Constants.ts
+++ b/src/Constants.ts
@@ -1,13 +1,8 @@
 import { RequestInit } from 'node-fetch';
 import siteJson from './sites.json';
-import SearchParameters from './structures/SearchParameters.js';
 import Site from './structures/Site';
 import SiteInfo from './structures/SiteInfo';

-export type searchFunction = {
-  search: (tags: string | string[], SearchArgs: SearchParameters) => any;
-};
-
 export interface SMap<V> {
   [key: string]: V;
 }
```

```diff
diff --git a/src/Utils.ts b/src/Utils.ts
index e7e57a3..e93818d 100644
--- a/src/Utils.ts
+++ b/src/Utils.ts
@@ -23,7 +23,7 @@ export function resolveSite (domain: string): string | null {
   domain = domain.toLowerCase();

   for (const site in sites) {
-    if (site === domain || (sites[site].aliases as string[]).includes(domain)) {
+    if (site === domain || sites[site].aliases.includes(domain)) {
       return site;
     }
   }
```

```diff
diff --git a/src/index.ts b/src/index.ts
index 70e929c..bb3f3bd 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -2,12 +2,11 @@ import { deprecate } from 'util';
 import Booru from './boorus/Booru';
 import Derpibooru from './boorus/Derpibooru';
 import XmlBooru from './boorus/XmlBooru';
-import { BooruError, searchFunction, sites, SMap } from './Constants';
+import { BooruError, sites, SMap } from './Constants';
 import Post from './structures/Post';
 import SearchParameters from './structures/SearchParameters';
 import SearchResults from './structures/SearchResults';
 import Site from './structures/Site';
-import SiteInfo from './structures/SiteInfo';
 import { resolveSite } from './Utils';

 const BooruTypes: any = {
@@ -15,7 +14,7 @@ const BooruTypes: any = {
   xml: XmlBooru,
 };

-const booruCache: SMap<SiteInfo & searchFunction> = {};
+const booruCache: SMap<any> = {};

 /**
  * Create a new booru, if special type, use that booru, else use default Booru
```

```diff
diff --git a/src/structures/SiteInfo.ts b/src/structures/SiteInfo.ts
index c35dcf5..7ff86b3 100644
--- a/src/structures/SiteInfo.ts
+++ b/src/structures/SiteInfo.ts
@@ -9,17 +9,17 @@ export default interface SiteInfo {
   /** The domain of the Site (the "google.com" part of "https://google.com/foo") */
   domain: string;
   /** The type of this site (json/xml/derpi) */
-  type?: string;
+  type: string;
   /** The aliases of this site */
-  aliases?: string[];
+  aliases: string[];
   /** If this site serves NSFW posts or not */
-  nsfw?: boolean;
+  nsfw: boolean;
   /** An object representing the api of this site */
-  api?: SiteApi;
+  api: SiteApi;
   /** The url query param to paginate on the site */
   paginate?: string;
   /** If the site supports `order:random` */
-  random?: boolean | string;
+  random: boolean | string;
   /** The url query param for tags */
   tagQuery?: string;
   /** If this site supports only http:// */
```

Reply:
Well that should finally be the last one then - addressed this in 7b6490a218bc6a9aded6fe2f70673334ddd7ccb1